clinica/iotools/converters/ukb_to_bids/ukb_utils.py

```diff
@@ -0,0 +1,386 @@
+    df_scan = df_clinical.filter(items=list(df_ref["scan"]))
+    return df_participants, df_session, df_scan
+
+
+def write_bids(
+    to: PathLike,
+    participants: DataFrame,
+    sessions: DataFrame,
+    scans: DataFrame,
+    dataset_directory: PathLike,
+) -> None:
+    from pathlib import Path
+
+    from fsspec.implementations.local import LocalFileSystem
+
+    from clinica.iotools.bids_dataset_description import BIDSDatasetDescription
+    from clinica.iotools.bids_utils import write_to_tsv
+
+    to = Path(to)
+    fs = LocalFileSystem(auto_mkdir=True)
+
+    participants = participants.droplevel(
+        ["session", "modality", "bids_filename"]
+    ).drop_duplicates()
+
+    # Ensure BIDS hierarchy is written first.
+    with fs.transaction:
+        with fs.open(to / "dataset_description.json", "w") as dataset_description_file:
+            BIDSDatasetDescription(name="UKB").write(to=dataset_description_file)
+
+        with fs.open(to / "participants.tsv", "w") as participant_file:
+            write_to_tsv(participants, participant_file)
+
+    for participant_id, data_frame in sessions.groupby(["participant_id"]):
+        session = data_frame.droplevel(
+            ["participant_id", "modality", "bids_filename"]
+        ).drop_duplicates()
+
+        session_filepath = to / participant_id / f"{participant_id}_sessions.tsv"
+        with fs.open(session_filepath, "w") as sessions_file:
+            write_to_tsv(session, sessions_file)
+    scans = scans.set_index(["bids_full_path"], verify_integrity=True)
+    for bids_full_path, metadata in scans.iterrows():
+        if metadata["modality_num"] != "20217" and metadata["modality_num"] != "20225":
+            copy_file_to_bids(
+                zipfile=str(dataset_directory) + "/" + metadata["source_zipfile"],
+                filenames=[metadata["source_filename"]] + metadata["sidecars"],
+                bids_path=to / bids_full_path,
+            )
+        else:
+            convert_dicom_to_nifti(
+                zipfiles=str(dataset_directory) + "/" + metadata["source_zipfile"],
+                bids_path=to / bids_full_path,
+            )
+    return
+
+
+def copy_file_to_bids(zipfile: str, filenames: List[str], bids_path: str) -> None:
+    """Install the requested files in the BIDS  dataset."""
+    import fsspec
+
+    fo = fsspec.open(zipfile)
+    fs = fsspec.filesystem("zip", fo=fo)
+    for filename in filenames:
+        if fs.exists(filename):
+            bids_path_extension = str(bids_path) + "." + (filename.split(".", 1)[1])
+            with fsspec.open(bids_path_extension, mode="wb") as f:
+                f.write(fs.cat(filename))
+
+
+def convert_dicom_to_nifti(zipfiles: str, bids_path: str) -> None:
+    """Install the requested files in the BIDS  dataset."""
+    import os
+    import subprocess
+    import tempfile
+    import zipfile
+
+    zf = zipfile.ZipFile(zipfiles)
+    try:
+        os.makedirs(bids_path.parent)
```

Comment:
The function signature says that `bids_path` is assumed to be a string while this is assuming a Path object without any cast or checks.